lib/puppet/util.rb

```diff
@@ -477,11 +477,11 @@ def uri_encode(path, opts = { :allow_fragment => false })
   end
   module_function :uri_encode
 
-  def safe_posix_fork(stdin=$stdin, stdout=$stdout, stderr=$stderr, &block)
+  def safe_posix_fork(stdin=STDIN, stdout=STDOUT, stderr=STDERR, &block)
     child_pid = Kernel.fork do
-      $stdin.reopen(stdin)
-      $stdout.reopen(stdout)
-      $stderr.reopen(stderr)
```

Comment:
I think we also want to set `$stdin`, `$stdout`, and `$stderr` to their global forms once this stuff has all been re-opened

Reply:
To add a bit more detail to what I said above:

Setting STDOUT will fix anything we `exec`, but any Ruby code run in a fork will still have `$stdout == STDERR` at this point unless we re-set `$stdout = STDOUT`